editoast/src/views/scenario/macro_nodes.rs

```diff
@@ -0,0 +1,635 @@
+    use crate::views::test_app::TestAppBuilder;
+
+    #[rstest]
+    async fn create() {
+        let app = TestAppBuilder::default_app();
+        let db_pool = app.db_pool();
+
+        let fixtures =
+            create_scenario_fixtures_set(&mut db_pool.get_ok(), "test_scenario_name").await;
+
+        let request = app
+            .post(&format!(
+                "/projects/{}/studies/{}/scenarios/{}/macro_nodes",
+                fixtures.project.id, fixtures.study.id, fixtures.scenario.id
+            ))
+            .json(&json!({
+                "position_x": 12,
+                "position_y": 51,
+                "full_name": "My super node",
+                "connection_time": 13,
+                "labels": ["",""],
+                "path_item_key":"->"
+            }));
+        let response: MacroNodeResponse =
+            app.fetch(request).assert_status(StatusCode::OK).json_into();
+
+        let node = MacroNode::retrieve_or_fail(&mut db_pool.get_ok(), response.id, || {
+            MacroNodeError::NotFound {
+                node_id: response.id,
+            }
+        })
+        .await
+        .expect("Failed to retrieve node");
+
+        assert_eq!(node.id, response.id);
+        assert_eq!(node.full_name, response.full_name);
+        assert_eq!(node.position_x, response.position_x);
+        assert_eq!(node.position_y, response.position_y);
+    }
+
+    #[rstest]
+    async fn update() {
+        let app = TestAppBuilder::default_app();
+        let db_pool = app.db_pool();
+        let fixtures = create_macro_node_fixtures_set(&mut db_pool.get_ok(), 1).await;
+
+        let request = app
+            .put(&format!(
+                "/projects/{}/studies/{}/scenarios/{}/macro_nodes/{}",
+                fixtures.project.id, fixtures.study.id, fixtures.scenario.id, fixtures.nodes[0].id
+            ))
+            .json(&json!({
+                "position_x": 4.0,
+                "position_y": 1.0,
+                "full_name": "My super node",
+                "connection_time": 13,
+                "labels": ["A","B"],
+                "path_item_key":"A->B"
+            }));
+        let response: MacroNodeResponse =
+            app.fetch(request).assert_status(StatusCode::OK).json_into();
+
+        let node = MacroNode::retrieve_or_fail(&mut db_pool.get_ok(), fixtures.nodes[0].id, || {
+            MacroNodeError::NotFound {
+                node_id: response.id,
+            }
+        })
+        .await
+        .expect("Failed to retrieve node");
+
+        let node_labels = node.labels.to_vec();
+        assert_eq!(Some("My super node".to_string()), node.full_name);
+        assert_eq!(4.0, node.position_x,);
+        assert_eq!(1.0, node.position_y);
+        assert_eq!("A".to_string(), node_labels[0]);
+        assert_eq!("B".to_string(), node_labels[1]);
+        assert_eq!(13, node.connection_time);
+        assert_eq!("A->B".to_string(), node.path_item_key);
+        assert_eq!(None, node.trigram);
```

Comment:
You can also use your `PartialEq` impl to `assert_eq!(fixture.node, node)`. `pretty_assertion` is useful for comparing structs as it prints a nice diff if the assert fails.